clients/drcachesim/tests/raw2trace_unit_tests.cpp

```diff
@@ -447,12 +458,158 @@ test_marker_placement(void *drcontext)
         check_entry(entries, idx, TRACE_TYPE_FOOTER, -1));
 }
 
+bool
+test_marker_delays(void *drcontext)
+{
+    // Our synthetic test first constructs a list of instructions to be encoded into
+    // a buffer for decoding by raw2trace.
+    instrlist_t *ilist = instrlist_create(drcontext);
+    // raw2trace doesn't like offsets of 0 so we shift with a nop.
+    instr_t *nop = XINST_CREATE_nop(drcontext);
+    // We test these scenarios:
+    // 1) Ensure that markers are delayed along with branches but timestamps and cpu
+    //    headers are not delayed along with branches.
+    // 2) Ensure that markers are not delayed across timestamp+cpu headers if there is no
+    //    branch also being delayed.
+    // 3) Ensure that markers are not delayed across window boundaries
+    // (TRACE_MARKER_TYPE_WINDOW_ID with a new id).
+    instr_t *move1 =
+        XINST_CREATE_move(drcontext, opnd_create_reg(REG1), opnd_create_reg(REG2));
+    instr_t *jmp = XINST_CREATE_jump(drcontext, opnd_create_instr(move1));
+    instr_t *move2 =
+        XINST_CREATE_move(drcontext, opnd_create_reg(REG1), opnd_create_reg(REG2));
+    instr_t *move3 =
+        XINST_CREATE_move(drcontext, opnd_create_reg(REG1), opnd_create_reg(REG2));
+    instr_t *move4 =
+        XINST_CREATE_move(drcontext, opnd_create_reg(REG1), opnd_create_reg(REG2));
+    instr_t *move5 =
+        XINST_CREATE_move(drcontext, opnd_create_reg(REG1), opnd_create_reg(REG2));
+    instrlist_append(ilist, nop);
+    // Block 1.
+    instrlist_append(ilist, move1);
+    instrlist_append(ilist, jmp);
+    // Block 2.
+    instrlist_append(ilist, move2);
+    instrlist_append(ilist, move3);
+    // Block 3.
+    instrlist_append(ilist, move4);
+    instrlist_append(ilist, move5);
+
+    size_t offs_nop = 0;
+    size_t offs_move1 = offs_nop + instr_length(drcontext, nop);
+    size_t offs_jmp = offs_move1 + instr_length(drcontext, move1);
+    size_t offs_move2 = offs_jmp + instr_length(drcontext, jmp);
+    size_t offs_move3 = offs_move2 + instr_length(drcontext, move2);
+    size_t offs_move4 = offs_move3 + instr_length(drcontext, move3);
+
+    // Now we synthesize our raw trace itself, including a valid header sequence.
+    std::vector<offline_entry_t> raw;
+    raw.push_back(make_header());
+    raw.push_back(make_tid());
+    raw.push_back(make_pid());
+    raw.push_back(make_line_size());
+    raw.push_back(make_block(offs_move1, 2));
```

Comment:
Comments here on which part belongs to which of tests 1, 2, and 3 would help.  So this is 1: the branch at the end of this block will be delayed until the next block is found: but it should cross the timestmap+cpu below, and carry the 3 func markers with it and not pass over those.  Please put something like this in a comment.